EVENTS.md

```diff
@@ -35,21 +35,85 @@ TODO: show event structure.
 
 TODO: contrast flattened/unflattened events
 
-### Standard Events in the SDK
+### Standard Events in the Cosmos-SDK
+The Cosmos-SDK emits events of type `sdk.EventTypeMessage` for every message that contains the `sender` (actor) and `module` field (which seems to be optional unfortunately)
+For example in the **staking** [msg_server.go](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/staking/keeper/msg_server.go#L106-L117):
+```go
+sdk.NewEvent(
+    types.EventTypeCreateValidator,
+    sdk.NewAttribute(types.AttributeKeyValidator, msg.ValidatorAddress),
+    sdk.NewAttribute(sdk.AttributeKeyAmount, msg.Value.Amount.String()),
+)
+```
+
+They also emit one or more other custom type event in the module that describe the operation executed with more details or metadata.
+```go
+sdk.NewEvent(
+  types.EventTypeCreateValidator,
+  sdk.NewAttribute(types.AttributeKeyValidator, msg.ValidatorAddress),
+  sdk.NewAttribute(sdk.AttributeKeyAmount, msg.Value.Amount.String()),
+)
+```
+
+Another example for a custom event is in the **bank** [send.go](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/bank/keeper/send.go#L117-L121),
+that shows a different set of metadata.
+```go
+sdk.NewEvent(
+    types.EventTypeTransfer,
+    sdk.NewAttribute(types.AttributeKeyRecipient, out.Address),
+    sdk.NewAttribute(sdk.AttributeKeyAmount, out.Coins.String()),
+),
+```
+and the `message` type but without the `module` this time. [See](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/bank/keeper/send.go#L117-L121) 
+```go
+sdk.NewEvent(
+    sdk.EventTypeMessage,
+    sdk.NewAttribute(types.AttributeKeySender, in.Address),
+),
+```
+The big difference between both modules is where the events are emitted. In the `message_server` or in the `keeper`. The "classic"
+way to use module functionality was by using the module's keeper. This may have driven the decision where to emit the events.
+
+The **distribution** module though has split the responsibility for the different event types better.
+In [`msg_server.go`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/distribution/keeper/msg_server.go#L42-L46)
+the `message` type is emitted:
+```go
+sdk.NewEvent(
+    sdk.EventTypeMessage,
+    sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+    sdk.NewAttribute(sdk.AttributeKeySender, msg.DelegatorAddress),
+),
+```
+While the custom event type is emitted in the [`keeper.go`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/distribution/keeper/keeper.go#L74-L77):
+```go
+sdk.NewEvent(
+    types.EventTypeSetWithdrawAddress,
+    sdk.NewAttribute(types.AttributeKeyWithdrawAddress, withdrawAddr.String()),
+),
+```
 
-TODO: what is added by the AnteHandlers (message.signer? auth?)
```

Comment:
This is emitted on every transaction - https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/auth/ante/sigverify.go#L103-L120 not so interesting for us but good to reference.

Also, it would make reading easier IMO if `sdk.AttributeKeyModule` was replaced by "module" or whatever. Anyone using events will need to find the strings for the client.

Reply:
And THIS is what I was thinking of when I mentioned AnteHandlers: https://github.com/cosmos/cosmos-sdk/blob/6888de1d86026c25197c1227dae3d7da4d41a441/baseapp/baseapp.go#L746-L748

```go
msgEvents := sdk.Events{
			sdk.NewEvent(sdk.EventTypeMessage, sdk.NewAttribute(sdk.AttributeKeyAction, eventMsgName)),
		}
		msgEvents = msgEvents.AppendEvents(msgResult.GetEvents())
```

Unfortunately the calculation of that string changed between 0.42.9 and 0.43.0 (for new message service modules):

0.42.9: `msgFqName = svcMsg.MethodName` https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/baseapp/baseapp.go#L709

0.43.0: `eventMsgName = sdk.MsgTypeURL(msg)` https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-rc3/baseapp/baseapp.go#L720

It uses a longer URL. This did break relayers